File: scripts/download_nces_data.R
# Download NCES Common Core of Data for High School Standardization
# This script downloads public school directory data from the Urban Institute Education Data API
# and NCES CCD for use in standardizing high school names

library(tidyverse)
library(httr)
library(jsonlite)

# Create data directory if it doesn't exist
dir.create("data/nces", recursive = TRUE, showWarnings = FALSE)

cat("Starting NCES data download...\n\n")

# ============================================================================
# METHOD 1: Urban Institute Education Data API
# ============================================================================
# This API provides easy access to CCD data with filtering capabilities
# Documentation: https://educationdata.urban.org/documentation/

download_urban_api <- function() {
  cat("Method 1: Attempting download via Urban Institute API...\n")

  # We'll download data for recent years to capture school information
  # School level codes: 1=Primary, 2=Middle, 3=High, 4=Other
  # We want high schools (3) and combined schools that include high school grades

  years <- c(2023, 2022, 2021)  # Most recent available years
  all_schools <- tibble()

  for (year in years) {
    cat(sprintf("  Downloading %d data...\n", year))

    # API endpoint for CCD school directory
    base_url <- sprintf("https://educationdata.urban.org/api/v1/schools/ccd/directory/%d/", year)

    # We'll need to paginate through results
    page <- 0
    year_schools <- tibble()

    repeat {
      # Build URL with pagination
      url <- paste0(base_url, "?page=", page)

      tryCatch({
        response <- GET(url)

        if (status_code(response) != 200) {
          cat(sprintf("    Error: HTTP %d\n", status_code(response)))
          break
        }

        data <- content(response, as = "text") %>% fromJSON()

        # Check if we got results
        if (is.null(data$results) || length(data$results) == 0) {
          break
        }

        # Append to year data
        page_data <- as_tibble(data$results)
        year_schools <- bind_rows(year_schools, page_data)

        cat(sprintf("    Page %d: %d schools (total: %d)\n",
                    page, nrow(page_data), nrow(year_schools)))

        # Check if there's a next page
        if (is.null(data$`next`) || data$`next` == "") {
          break
        }

        page <- page + 1

        # Rate limiting - be nice to the API
        Sys.sleep(0.5)

      }, error = function(e) {
        cat(sprintf("    Error on page %d: %s\n", page, e$message))
        break
      })
    }

    if (nrow(year_schools) > 0) {
      year_schools$data_year <- year
      all_schools <- bind_rows(all_schools, year_schools)
      cat(sprintf("  Year %d complete: %d schools\n", year, nrow(year_schools)))
    }
  }

  if (nrow(all_schools) > 0) {
    # Save raw data
    write_csv(all_schools, "data/nces/urban_api_raw.csv")
    cat(sprintf("\nDownloaded %d school records from Urban Institute API\n", nrow(all_schools)))
    return(all_schools)
  } else {
    cat("\nFailed to download from Urban Institute API\n")
    return(NULL)
  }
}

# ============================================================================
# METHOD 2: Direct NCES Download (manual backup)
# ============================================================================
# If API fails, provide instructions for manual download

print_manual_instructions <- function() {
  cat("\n============================================================\n")
  cat("MANUAL DOWNLOAD INSTRUCTIONS\n")
  cat("============================================================\n\n")
  cat("If automatic download fails, please manually download:\n\n")
  cat("1. Go to: https://nces.ed.gov/ccd/files.asp\n")
  cat("2. Find '2023-24 Public Elementary/Secondary School Universe Survey'\n")
  cat("3. Download the CSV file (usually named ccd_sch_*.csv)\n")
  cat("4. Save it to: data/nces/nces_schools_manual.csv\n\n")
  cat("Alternative:\n")
  cat("1. Go to: https://nces.ed.gov/ccd/pubschuniv.asp\n")
  cat("2. Download the most recent school universe survey data\n")
  cat("3. Extract and save the school-level CSV to data/nces/\n\n")
  cat("============================================================\n\n")
}

# ============================================================================
# Filter for High Schools
# ============================================================================

filter_high_schools <- function(schools_data) {
  cat("\nFiltering for high schools...\n")

  # Identify columns that might indicate school level
  # Different data sources use different column names
  level_cols <- c("school_level", "level", "lowest_grade", "highest_grade")
  grade_cols <- c("lowest_grade", "highest_grade", "lowest_grade_offered",
                  "highest_grade_offered")

  # Filter criteria:
  # - School level = 3 (High) or 4 (Other - might include alternative high schools)
  # - OR highest grade >= 9 (includes high school grades)
  # - School status = Open/Operating (not closed)

  hs_data <- schools_data %>%
    filter(
      # Filter for open schools
      if("school_status" %in% names(.)) school_status == 1 else TRUE,
      # Filter for schools with high school grades
      if("school_level" %in% names(.)) {
        school_level %in% c(3, 4) | is.na(school_level)
      } else {
        TRUE
      },
      if("highest_grade" %in% names(.)) {
        highest_grade >= 9 | is.na(highest_grade)
      } else if("highest_grade_offered" %in% names(.)) {
        highest_grade_offered >= 9 | is.na(highest_grade_offered)
      } else {
        TRUE
      }
    )

  cat(sprintf("Filtered to %d high schools from %d total schools\n",
              nrow(hs_data), nrow(schools_data)))

  return(hs_data)
}

# ============================================================================
# Standardize Column Names
# ============================================================================

standardize_columns <- function(schools_data) {
  cat("\nStandardizing column names...\n")

  # Map common column name variations to standard names
  column_mapping <- c(
    "ncessch" = "nces_school_id",
    "ncessch_num" = "nces_school_id",
    "school_name" = "school_name",
    "schnam" = "school_name",
    "sch_name" = "school_name",
    "lea_name" = "district_name",
    "leanm" = "district_name",
    "leaid" = "nces_district_id",
    "state_location" = "state",
    "state_leaid" = "state",
    "lstate" = "state",
    "lcity" = "city",
    "city_location" = "city",
    "lstree" = "street",
    "location_address" = "street",
    "lzip" = "zip",
    "phone" = "phone",
    "school_level" = "school_level",
    "lowest_grade" = "lowest_grade",
    "highest_grade" = "highest_grade"
  )

  # Rename columns that exist
  for (old_name in names(column_mapping)) {
    new_name <- column_mapping[[old_name]]
    if (old_name %in% names(schools_data) && !new_name %in% names(schools_data)) {
      schools_data <- schools_data %>%
        rename(!!new_name := !!old_name)
    }
  }

  cat("Column standardization complete\n")

  return(schools_data)
}

# ============================================================================
# Main Execution
# ============================================================================

# Try Urban Institute API first
schools_data <- download_urban_api()

# If API download worked, process the data
if (!is.null(schools_data) && nrow(schools_data) > 0) {

  # Standardize column names
  schools_data <- standardize_columns(schools_data)

  # Filter for high schools
  hs_data <- filter_high_schools(schools_data)

  # Save processed high school data
  write_csv(hs_data, "data/nces/high_schools_all_years.csv")

  # Also create a deduplicated version (most recent data per school)
  hs_dedup <- hs_data %>%
    arrange(nces_school_id, desc(data_year)) %>%
    distinct(nces_school_id, .keep_all = TRUE)

  write_csv(hs_dedup, "data/nces/high_schools_current.csv")

  # Print summary statistics
  cat("\n============================================================\n")
  cat("DOWNLOAD COMPLETE\n")
  cat("============================================================\n\n")
  cat(sprintf("Total high schools: %d\n", nrow(hs_dedup)))
  cat(sprintf("States covered: %d\n", n_distinct(hs_dedup$state)))
  cat("\nFiles created:\n")
  cat("  - data/nces/urban_api_raw.csv (raw API data)\n")
  cat("  - data/nces/high_schools_all_years.csv (all years)\n")
  cat("  - data/nces/high_schools_current.csv (deduplicated)\n\n")

  # Show sample of data
  cat("Sample of downloaded data:\n")
  hs_dedup %>%
    select(any_of(c("nces_school_id", "school_name", "city", "state"))) %>%
    head(10) %>%
    print()

} else {
  # API failed, show manual instructions
  print_manual_instructions()
  cat("\nScript will continue with next steps once data is available.\n")
}

cat("\nNECS data download script complete.\n")

File: scripts/download_nces_data.py
#!/usr/bin/env python3
"""
Download NCES Common Core of Data for High School Standardization
This script downloads public school directory data from the Urban Institute Education Data API
"""

import requests
import pandas as pd
import time
import os
from pathlib import Path

# Create data directory
DATA_DIR = Path("data/nces")
DATA_DIR.mkdir(parents=True, exist_ok=True)

print("Starting NCES data download...\n")

def download_from_urban_api(year=2022):
    """
    Download school directory data from Urban Institute Education Data API
    Note: 2022 is used as it's widely available; 2023 may have limited data
    """
    print(f"Downloading data for year {year} from Urban Institute API...")

    base_url = f"https://educationdata.urban.org/api/v1/schools/ccd/directory/{year}/"

    all_schools = []
    page = 0
    max_pages = 100  # Safety limit

    while page < max_pages:
        try:
            # Build URL with parameters
            params = {
                'page': page,
                'per_page': 5000  # Maximum records per page
            }

            print(f"  Fetching page {page}...", end='', flush=True)

            response = requests.get(base_url, params=params, timeout=30)

            if response.status_code != 200:
                print(f" Error: HTTP {response.status_code}")
                break

            data = response.json()

            # Check if we got results
            if 'results' not in data or len(data['results']) == 0:
                print(" No more results")
                break

            all_schools.extend(data['results'])
            print(f" Got {len(data['results'])} schools (total: {len(all_schools)})")

            # Check if there's more data
            if 'next' not in data or not data['next']:
                print("  Reached last page")
                break

            page += 1

            # Rate limiting
            time.sleep(0.3)

        except Exception as e:
            print(f" Error: {e}")
            break

    if all_schools:
        df = pd.DataFrame(all_schools)
        print(f"\nTotal schools downloaded: {len(df)}")
        return df
    else:
        print("\nNo data downloaded")
        return None


def filter_high_schools(df):
    """Filter for high schools only"""
    print("\nFiltering for high schools...")

    # Filter criteria:
    # - school_level == 3 (High School) or 4 (Other)
    # - OR highest_grade >= 9
    # - school_status == 1 (Open)

    initial_count = len(df)

    # Filter for open schools
    if 'school_status' in df.columns:
        df = df[df['school_status'] == 1]

    # Filter for high schools
    # School level: 1=Primary, 2=Middle, 3=High, 4=Other
    high_school_df = df[
        ((df.get('school_level', 0).isin([3, 4])) |
         (df.get('highest_grade', 0) >= 9))
    ].copy()

    print(f"Filtered from {initial_count} to {len(high_school_df)} high schools")

    return high_school_df


def standardize_columns(df):
    """Standardize column names for easier use"""
    print("\nStandardizing column names...")

    # Column mapping
    column_map = {
        'ncessch': 'nces_school_id',
        'school_name': 'school_name',
        'lea_name': 'district_name',
        'leaid': 'nces_district_id',
        'state_location': 'state',
        'city_location': 'city',
        'zip_location': 'zip',
        'street_location': 'street',
        'phone': 'phone',
        'school_level': 'school_level',
        'lowest_grade': 'lowest_grade',
        'highest_grade': 'highest_grade',
        'enrollment': 'enrollment',
        'county_name': 'county',
        'county_code': 'county_code',
        'fips': 'state_fips'
    }

    # Rename columns that exist
    df = df.rename(columns={k: v for k, v in column_map.items() if k in df.columns})

    print("Standardization complete")
    return df


def download_state_by_state(year=2022):
    """
    Alternative method: Download state by state
    This is more reliable than trying to paginate through all results
    """
    print(f"Downloading state-by-state data for year {year}...")

    # FIPS codes for all US states and DC
    state_fips = {
        1: 'AL', 2: 'AK', 4: 'AZ', 5: 'AR', 6: 'CA', 8: 'CO', 9: 'CT', 10: 'DE',
        11: 'DC', 12: 'FL', 13: 'GA', 15: 'HI', 16: 'ID', 17: 'IL', 18: 'IN',
        19: 'IA', 20: 'KS', 21: 'KY', 22: 'LA', 23: 'ME', 24: 'MD', 25: 'MA',
        26: 'MI', 27: 'MN', 28: 'MS', 29: 'MO', 30: 'MT', 31: 'NE', 32: 'NV',
        33: 'NH', 34: 'NJ', 35: 'NM', 36: 'NY', 37: 'NC', 38: 'ND', 39: 'OH',
        40: 'OK', 41: 'OR', 42: 'PA', 44: 'RI', 45: 'SC', 46: 'SD', 47: 'TN',
        48: 'TX', 49: 'UT', 50: 'VT', 51: 'VA', 53: 'WA', 54: 'WV', 55: 'WI',
        56: 'WY'
    }

    all_schools = []

    for fips, state_abbr in state_fips.items():
        try:
            print(f"  Downloading {state_abbr} (FIPS {fips})...", end='', flush=True)

            url = f"https://educationdata.urban.org/api/v1/schools/ccd/directory/{year}/"
            params = {'fips': fips}

            response = requests.get(url, params=params, timeout=30)

            if response.status_code != 200:
                print(f" Error: HTTP {response.status_code}")
                continue

            data = response.json()

            if 'results' in data and data['results']:
                state_schools = data['results']
                all_schools.extend(state_schools)
                print(f" {len(state_schools)} schools")
            else:
                print(" No data")

            # Rate limiting
            time.sleep(0.2)

        except Exception as e:
            print(f" Error: {e}")
            continue

    if all_schools:
        df = pd.DataFrame(all_schools)
        print(f"\nTotal schools downloaded: {len(df)}")
        return df
    else:
        print("\nNo data downloaded")
        return None


def main():
    """Main execution"""

    # Try state-by-state download first (more reliable)
    print("Method 1: State-by-state download\n")
    df = download_state_by_state(year=2022)

    if df is None or len(df) == 0:
        print("\nMethod 1 failed. Trying Method 2: Full download\n")
        df = download_from_urban_api(year=2022)

    if df is None or len(df) == 0:
        print("\n" + "="*60)
        print("DOWNLOAD FAILED")
        print("="*60)
        print("\nPlease manually download NCES data from:")
        print("https://nces.ed.gov/ccd/files.asp")
        print("\nSave as: data/nces/nces_schools_manual.csv")
        return

    # Save raw data
    raw_file = DATA_DIR / "urban_api_raw.csv"
    df.to_csv(raw_file, index=False)
    print(f"\nSaved raw data: {raw_file}")

    # Standardize columns
    df = standardize_columns(df)

    # Filter for high schools
    hs_df = filter_high_schools(df)

    # Save high schools
    hs_file = DATA_DIR / "high_schools_current.csv"
    hs_df.to_csv(hs_file, index=False)
    print(f"Saved high school data: {hs_file}")

    # Create summary
    print("\n" + "="*60)
    print("DOWNLOAD COMPLETE")
    print("="*60)
    print(f"\nTotal high schools: {len(hs_df)}")

    if 'state' in hs_df.columns:
        print(f"States covered: {hs_df['state'].nunique()}")
        print(f"\nTop 10 states by school count:")
        print(hs_df['state'].value_counts().head(10))

    print("\nFiles created:")
    print(f"  - {raw_file}")
    print(f"  - {hs_file}")

    # Show sample
    print("\nSample of data:")
    cols_to_show = ['nces_school_id', 'school_name', 'city', 'state']
    available_cols = [c for c in cols_to_show if c in hs_df.columns]
    if available_cols:
        print(hs_df[available_cols].head(10).to_string(index=False))

    print("\nNECS data download complete!")


if __name__ == "__main__":
    main()
